#72: Added song synchronization between host and clients

MatanRad opened the pull request:
Added synchronization of soundtrack between the host and the clients.
Since the soundtrack is not saved in the replay or the saves, I implemented it using a new command.
In multiplayer, the clients (not the host) do not choose or automatically play any non-forced music and the host sends an update for each song that is starting to play. When the clients receive the update, the forcibly play the songs.

notfood commented:
Hesitant to include this one, as with random things, this should be in sync already. 

What it does is, makes the host authoritative of which song was picked at random instead of using the same random seed to pick the song. Shouldn't we be fixing the rng instead?

MatanRad commented:
> Hesitant to include this one, as with random things, this should be in sync already.
> 
> What it does is, makes the host authoritative of which song was picked at random instead of using the same random seed to pick the song. Shouldn't we be fixing the rng instead?

I talked about this briefly with Nebual on the Discord server, since the state of the MusicManagerPlay is not saved in the replay or the save data, it will be hard to make this remain in sync. 
For example, if a player joins mid-track (which is almost always the case) or if a desync occurs then it will be much harder to synchronize them.

Since it is kinda detached from the rest of the game-state, I think it might be better to just make the host authoritative of the order of the songs.

MatanRad commented:
I just want to make sure I understood the problem you've mentioned correctly since I didn't think about the simulation phase.
Is that correct?
(That is what I understood at least. Practically it won't start a song since the patch prevents the client from playing non-forced songs)

I'm surprised that you kept running into that exception, I hosted it a bunch of times today and tested it with a friend. Currently I'm unable to check for that exception since I am away until Wednesday.
I fixed the remaining CR notes and added the fix you mentioned for the exception but unfortunately couldn't sit down and properly check it.
I'll perform more comprehensive tests as soon as I am back :)

Zetrith commented:
Doing this in a bug-free way would probably be too much work for too little gain. It might get revisited in the future.
